### PR TITLE
[inductor] [cpp] fix accuracy when template_buffer has users other than the epilogue nodes

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133073

This PR fixes the accuracy issues when template_buffer has users other than the epilogue nodes. This will fix the accuracy failure of the below models using max-autotune:

- MobileBertForMaskedLM
- MobileBertForQuestionAnswering
- convnext_base
- swin_base_patch4_window7_224

## Issue 1:
Previously we always add `template_buffer` as an alias of `Y`. In case the `template_buffer` has users other than the epilogue nodes, we shouldn't set it as an alias of `Y`. This PR adds the check in such case.

Wrong code before the fix where `tmp4` and `tmp9` are both stored to `Y` while we need 2 different buffers for them since `tmp4` will be used by nodes other than the epilogue node:
```cpp
Y[static_cast<long>(n_start + x1 + (32L*m_start) + (32L*x0))] = tmp4; // tmp4 is the output of the template
Y[static_cast<long>(n_start + x1 + (32L*m_start) + (32L*x0))] = tmp9; // tmp9 is the output of the epilogue node
```

Correct code after the fix:
```cpp
out_ptr2[static_cast<long>(n_start + x1 + (32L*m_start) + (32L*x0))] = tmp4;
Y[static_cast<long>(n_start + x1 + (32L*m_start) + (32L*x0))] = tmp9;
```

## Issue 2:
When fixing the above issue, we found that there's correctness issue when `bias` is `False`. The root cause is that in the case where `bias` is `False`, the `template_buffer` has users other than the epilogue nodes and the GEMM output buffer is localized, we need to add an extra copy epilogue to ensure that the GEMM output (a local buffer) is stored to the `template_buffer` that will be used later by other nodes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jgong5 — CHANGES_REQUESTED** [review]
Perhaps we can reuse the `decide_inplace_update` and existing inplace buffer handling logic?

**Comment by chunyuan-w** [comment]
> Perhaps we can reuse the `decide_inplace_update` and existing inplace buffer handling logic?

As synced offline, I've moved the `can_alias_template_with_kernel_output` check to `cpp.py` and directly pass the output flag of this function to the cpp gemm template.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 1]
```diff
@@ -700,9 +716,62 @@ def _bias_add_epilogue(buf):
...
+
+        if need_copy_from_local_to_global_buffer_epilogue(
+            use_local_acc, can_alias_template_with_kernel_output, epilogue_creators
+        ):
+
+            def copy_from_local_to_global_buffer_epilogue(input_buffer: ir.Buffer):
+                dtype = self.layout.dtype
+                input_loader = input_buffer.make_loader()
+
+                def copy_innter(index):
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:761]: nit: copy_inner

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -653,6 +653,7 @@ def render(  # type: ignore[override,return]
         self,
         kernel: CppTemplateKernel,
         template_buffer_node: Optional[ir.CppTemplateBuffer] = None,
+        flag_can_alias_template_with_kernel_output: Optional[bool] = None,
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:656]: how about "template_buffer_has_other_users"?

File: torch/_inductor/codegen/cpp_template_kernel.py [hunk 3]
```diff
@@ -324,7 +324,12 @@ def __init__(
         input_nodes: List[ir.Buffer],
         layout: ir.Layout,
         make_kernel_render: Callable[
-            [ir.CppTemplateBuffer, Optional[List[ir.IRNode]]], str
+            [
+                ir.CppTemplateBuffer,
+                bool,
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_template_kernel.py:329]: optional or not?

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -4230,7 +4230,37 @@ def codegen_template(
         assert all(
             isinstance(n, ir.ComputedBuffer) for n in epilogue_ir_nodes
         ), "Epilogue nodes must all be instances of ir.ComputedBuffer"
-        kernel, render = ctb.make_kernel_render(ctb, epilogue_nodes=epilogue_ir_nodes)
+
+        def can_alias_template_with_kernel_output(
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:4234]: how about "template_buffer_has_other_users"?

File: torch/_inductor/codegen/cpp.py [hunk 5]
```diff
@@ -4230,7 +4230,37 @@ def codegen_template(
...
+            template_buffer, outputs_by_name, epilogue_nodes
+        ):
+            # If there's no epilogue_nodes, template_buffer the and kernel output buffer
+            # are the same, thus return True here.
+            if not epilogue_nodes:
+                return True
+
+            assert template_buffer.get_name() in outputs_by_name
+            users = outputs_by_name[template_buffer.get_name()].users
+            for user in users:
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:4244]: perhaps using `all(...)` is simpler?

**Review by jgong5 — APPROVED** [review]
LGTM overall. Some small nits on the naming, typo etc.

File: torch/_inductor/codegen/cpp_template_kernel.py (hunk 3 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_template_kernel.py:329]: Optional in the `render` function but not optional here

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 1 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Fixed the typo.

File: torch/_inductor/codegen/cpp.py (hunk 5 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp.py]: Updated

File: torch/_inductor/codegen/cpp.py (hunk 4 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp.py]: Renamed

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 2 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Renamed

**Pull request closed without merging**